### DISCUSSION TITLE
VRS 2.0.0 Trial Use Ballot

### DISCUSSION THREAD

**Discussion by ahwagner** [ga4gh/vrs#533] [Announcements]
The first ballot release for the VRS specification, this ballot covers models under consideration for advancement to trial use in v2.0.0, including:

- Expression
- Allele
- CisPhasedBlock
- CopyNumberCount
- CopyNumberChange
- SequenceLocation
- SequenceReference
- ReferenceLengthExpression
- LengthExpression
- LiteralSequenceExpression
- Range
- Residue
- SequenceString
- Adjacency
- Terminal
- DerivativeMolecule
- TraversalBlock

<hr /><em>This discussion was created from the release <a href='https://github.com/ga4gh/vrs/releases/tag/2.0.0-ballot.2024-08.1'>VRS 2.0.0 Trial Use Ballot</a>.</em>

**Comment by rrfreimuth** [comment]
Thanks to everyone for putting so much effort into this release! Great job!

Comments and questions:

- Please add a figure showing the class hierarchy to illustrate inheritance. This would be especially helpful for an abstract class like Location, which is listed as a datatype for several attributes but is not included in the documentation so it is difficult to know which child classes are valid for a given attribute.
- Please include links to examples from within concept/model pages, so it is easier to see instance examples of each class
- Sequence Reference appears to support only refget accession numbers. How would other types of accession numbers be represented? (as a CURIE in .id?) Is it possible to specify more than one accession number?
- Why is SequenceLocation.sequence not of type LiteralSequenceExpression?
- SequenceLocation should require either sequenceReference or sequence to be populated
- The links to type IRI (referencing https://vrs.ga4gh.org/ga4gh/schema/gks-common/1.x/data-types/json/IRI) are broken
- The relationship between Allele and Derivative Molecule should be defined. They both represent contiguous sequences and it may be possible to represent at least some Derivative Molecules as Alleles. A normalization algorithm may be needed.
- Are all members of a CisPhasedBlock defined to be in the forward orientation?
- Adjacency: Please add an example that illustrates "increasing position" adjacent to "increasing position"
- Derivative Molecules: Is it required that sequential elements form a contiguous path? For example, if a DerMol contains 2 adjacencies (between two Terminii), should there be structural consistency (sequence, location, orientation) between the first adjoinedSequence of the second Adjacency and the second adjoinedSequence of the first Adjacency?
- How is Terminus different from Location?
- I am not sure Terminus should be a type of Molecular Variation (it seems more like a utility datatype)
- The expressions attribute is inherited by several classes but its use in those classes is not apparent (e.g., DerivativeMolecule, CopyNumberChange)
- Why is the alternativeLabels attribute needed, rather than simply expanding the cardinality of the label attribute to 0..m?

**Reply by ahwagner** [reply]
re: bullet 3 above ☝️ 
> Sequence Reference appears to support only refget accession numbers. How would other types of accession numbers be represented? (as a CURIE in .id?) Is it possible to specify more than one accession number?

@larrybabb suggests this is an opportunity to add a FHIR-like `Identifier` class and `Identifiers` property to `Entity`. Target for 2.1 release unless blocking for an implementation.

#563

**Reply by larrybabb** [reply]
Bob F mentioned that FHIR considers sequence accessions to be more like Codings (CodeableConcepts) rather than Identifiers.

**Reply by ahwagner** [reply]
> Please add a figure showing the class hierarchy to illustrate inheritance. This would be especially helpful for an abstract class like Location, which is listed as a datatype for several attributes but is not included in the documentation so it is difficult to know which child classes are valid for a given attribute.

Tracked in #561.

**Reply by ahwagner** [reply]
> Please include links to examples from within concept/model pages, so it is easier to see instance examples of each class

Tracked in #562.

**Reply by ahwagner** [reply]
> Why is SequenceLocation.sequence not of type LiteralSequenceExpression?

There is no need to differentiate between sequence expression types for this property, so we use the `SequenceString` class that is also used by `LiteralSequenceExpression` and reduce message overhead.

**Reply by ahwagner** [reply]
> SequenceLocation should require either sequenceReference or sequence to be populated

No. `Sequence` is just descriptive of the content described by the SequenceLocation, it is not definitive like a sequenceReference is. SequenceReference is optional due to its use in CisPhasedBlock Alleles. We recognize that the documentation needs to be extended to discuss this nuance and we are taking that issue on in issue #564 .

**Reply by ahwagner** [reply]
> The links to type IRI (referencing https://vrs.ga4gh.org/ga4gh/schema/gks-common/1.x/data-types/json/IRI) are broken

We weren't able to find this in the VRS documentation. Please provide a page URL on which this link is broken for us to address.

**Reply by ahwagner** [reply]
> The relationship between Allele and Derivative Molecule should be defined. They both represent contiguous sequences and it may be possible to represent at least some Derivative Molecules as Alleles. A normalization algorithm may be needed.

Derivative Molecule is a draft class, and we are not advancing it to Trial Use in 2.0; examples of the issue you are raising would be welcomed to help clarify if and what relationship between these classes needs to be defined.

We recognize that some _Adjacencies_ can represent the same end sequence state as _Allele_. We are not providing guidance on when implementations should use one or the other; it would be premature for us to do so at this stage. If this became an issue in real-world implementations, we would revisit this point.

**Reply by ahwagner** [reply]
> Are all members of a CisPhasedBlock defined to be in the forward orientation?

Yes, all `Variation` in VRS is defined in forward orientation on sequences. Orientation will be discussed in further detail when we are considering Derivative Molecule for Trial Use.

**Reply by ahwagner** [reply]
> Adjacency: Please add an example that illustrates "increasing position" adjacent to "increasing position"

This is not fundamentally different from the existing example with a reverse complement partner. @rrfreimuth what information is missing or unclear that another such example figure would be informative?

**Reply by ahwagner** [reply]
> Derivative Molecules: Is it required that sequential elements form a contiguous path? For example, if a DerMol contains 2 adjacencies (between two Terminii), should there be structural consistency (sequence, location, orientation) between the first adjoinedSequence of the second Adjacency and the second adjoinedSequence of the first Adjacency?

Yes, this is the intent of that class.

**Reply by ahwagner** [reply]
> How is Terminus different from Location?

A terminus is the explicit end of a physical molecule at a Location. The Location defines the coordinate where this occurs. The SV SIG recommended use of `Adjacency` for this, but the semantics were fuzzy so we created a new class, `Terminus`.

**Reply by ahwagner** [reply]
> I am not sure Terminus should be a type of Molecular Variation (it seems more like a utility datatype)

@rrfreimuth I would like to hear your rationale for this; please open a Discussion with additional detail to help motivate why a change is needed here. We believe that both Adjacency and Terminus objects share an abstract (undefined) parent concept, and so either both should be Mol Var or both something else.

**Reply by ahwagner** [reply]
> The expressions attribute is inherited by several classes but its use in those classes is not apparent (e.g., DerivativeMolecule, CopyNumberChange)

If or when community nomenclatures for these concepts exist, `expression` is useful for representing those nomenclatures as descriptive metadata. For example, one can envision use of ISCN representations for some DerivativeMolecules.

**Reply by ahwagner** [reply]
> Why is the alternativeLabels attribute needed, rather than simply expanding the cardinality of the label attribute to 0..m?

The label under the `label` attribute is the primary label for display.

**Comment by Mrinal-Thomas-Epic** [comment]
I didn't notice it until today's discussion, but why are Mappings no longer supported on VRS variation? If an implementation is using the `id` field for their own identifier, how would they attach ClinGen Allele Registry CA IDs or dbSNP rs IDs (or even a VRS 1.3 ID) to an Allele without a mappings concept?

**Reply by larrybabb** [reply]
We(Matt and I) made a change to only include mappings on `DomainEntity`.  Alex had originally created a special class called `MappableEntity` which existed in the ValueObject hierarchy. We thought it would be fine to have the VRS Variation and Location classes use `DomainEntity` as it's parent, but Alex pointed out later this was not logical (see Alex for logic).  In the end the mappings got dropped/lost from ValueObjects.

The question is whether FHIR `identifiers` can be handled by `Mappings` and if so, should we simply add this as an optional field at the `Entity` level?  I think `Mappings` were mainly introduced to handle `xrefs` since they provide the `relation` feature.  

In the end, we may need to have an `identifiers` attribute that is more of an "exact" mapping and useful for listing identifiers (not just the concept `id` within the message).  identifiers would presumably always be exact mappings and follow the FHIR definition for the concept.

CodeableConcept would then be the "thing" that is replaced by `Mappings` I suppose.

Lots to discuss.

**Reply by larrybabb** [reply]
Update on this discussion.  @ahwagner @mbrush and I are working out a draft/trial use plan for Mappings and DomainEntities as I write. We will be picking a path forward with the intent of allowing developers to establish the first steps. We will reserve the right to refactor this DomainEntity area of the model in a future minor or major release of the specification.
For now, we have agreed that DomainEntities belong only up in the VA-spec to relate to profiles that may need them.
